File: content/10-tries/implement_trie.py
"""
Problem: Implement Trie (Prefix Tree)
LeetCode: #208 (Medium) | NeetCode 150 | Blind 75
"""

class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    """
    Time Complexity:
      - insert: O(L) where L is key length
      - search: O(L)
      - startsWith: O(L)
    Space Complexity: O(T) where T is total number of characters inserted
    """
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root
        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()
            curr = curr.children[char]
        curr.is_end_of_word = True

    def search(self, word: str) -> bool:
        curr = self.root
        for char in word:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return curr.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        curr = self.root
        for char in prefix:
            if char not in curr.children:
                return False
            curr = curr.children[char]
        return True

if __name__ == "__main__":
    trie = Trie()
    trie.insert("apple")
    assert trie.search("apple") == True
    assert trie.search("app") == False
    assert trie.starts_with("app") == True
    trie.insert("app")
    assert trie.search("app") == True
    print("Python: Trie passed tests!")

File: content/10-tries/implement_trie.js
/**
 * Problem: Implement Trie (Prefix Tree)
 * LeetCode: #208 (Medium) | NeetCode 150 | Blind 75
 */

class TrieNode {
    constructor() {
        self.children = {};
        self.isEndOfWord = false;
    }
}

class Trie {
    /**
     * Time Complexity: O(L) for insert, search, and startsWith
     * Space Complexity: O(T)
     */
    constructor() {
        this.root = {};
    }

    insert(word) {
        let curr = this.root;
        for (const char of word) {
            if (!curr[char]) {
                curr[char] = {};
            }
            curr = curr[char];
        }
        curr.isEndOfWord = true;
    }

    search(word) {
        let curr = this.root;
        for (const char of word) {
            if (!curr[char]) {
                return false;
            }
            curr = curr[char];
        }
        return curr.isEndOfWord === true;
    }

    startsWith(prefix) {
        let curr = this.root;
        for (const char of prefix) {
            if (!curr[char]) {
                return false;
            }
            curr = curr[char];
        }
        return true;
    }
}

if (require.main === module) {
    const assert = require('assert');
    const trie = new Trie();
    trie.insert("apple");
    assert.strictEqual(trie.search("apple"), true);
    assert.strictEqual(trie.search("app"), false);
    assert.strictEqual(trie.startsWith("app"), true);
    trie.insert("app");
    assert.strictEqual(trie.search("app"), true);
    console.log("JS: Trie passed tests!");
}
